code.js

```diff
@@ -1,3 +1,41 @@
 function tsp_ls(distance_matrix) {
-    return -1;
+  let n = distance_matrix.length;
+  let memo = new Set();
+  let minRoute = Array.from({ length: n }, (v, i) => i).sort((a, b) => 0.5 - Math.random());
+  let minCost = calcCost(minRoute, distance_matrix);
+  let improving = true;
+  while(improving) {
+    let routePath = [...minRoute]; // Get a deep copy of min route
+    let r = Math.floor(Math.random() * (n - 1)) + 1; // Random Number in the range 1 to n - 1
+    let l = Math.floor(Math.random() * (n - r - 1)) + (r + 1); // Random number in the range r + 1 to n
+    swap(routePath, r, l);
+    // If route was already computed don't recompute it
+    let routeKey = routePath.join(",");
```

Comment:
Is that actually worth it? Computing the length of the path is very quick.

Reply:
I may have misread the question. I remade it without memoization.